gem/optimise.py

```diff
@@ -230,13 +230,47 @@ def delta_elimination(sum_indices, factors):
     return sum_indices, [e for e in factors if e != one]
 
 
+def associate_product(factors):
+    """Apply associativity rules to construct an operation-minimal product tree.
+
+    For best performance give factors that have different set of free indices.
+    """
```

Comment:
> For best performance give factors that have different set of free indices.

What this really means is that if you have multiple factors with the same set of free indices, then you are wasting optimisation time.